sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/NettyHttpClient.java

```diff
@@ -0,0 +1,253 @@
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoopGroup;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpDecoderConfig;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpHeadersFactory;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.proxy.HttpProxyHandler;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.stream.ChunkedInput;
+import io.netty.handler.stream.ChunkedNioFile;
+import io.netty.handler.stream.ChunkedStream;
+import io.netty.handler.stream.ChunkedWriteHandler;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.nio.channels.FileChannel;
+import java.nio.file.StandardOpenOption;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.netty.handler.codec.http.DefaultHttpHeadersFactory.trailersFactory;
+
+/**
+ * HttpClient implementation using synchronous Netty operations.
+ */
+class NettyHttpClient implements HttpClient {
+    private static final ClientLogger LOGGER = new ClientLogger(NettyHttpClient.class);
+
+    private final Bootstrap bootstrap;
+    private final SslContext sslContext;
+    private final ProxyOptions proxyOptions;
+    private final long readTimeoutMillis;
+    private final long responseTimeoutMillis;
+    private final long writeTimeoutMillis;
+
+    NettyHttpClient(Bootstrap bootstrap, SslContext sslContext, ProxyOptions proxyOptions, long readTimeoutMillis,
+        long responseTimeoutMillis, long writeTimeoutMillis) {
+        this.bootstrap = bootstrap;
+        this.sslContext = sslContext;
+        this.proxyOptions = proxyOptions;
+        this.readTimeoutMillis = readTimeoutMillis;
+        this.responseTimeoutMillis = responseTimeoutMillis;
+        this.writeTimeoutMillis = writeTimeoutMillis;
+    }
+
+    // For testing.
+    ProxyOptions getProxyOptions() {
+        return proxyOptions;
+    }
+
+    Bootstrap getBootstrap() {
+        return bootstrap;
+    }
+
+    @Override
+    public Response<BinaryData> send(HttpRequest request) throws IOException {
+        URI uri = request.getUri();
+        String host = uri.getHost();
+        int port = uri.getPort() == -1 ? ("https".equalsIgnoreCase(uri.getScheme()) ? 443 : 80) : uri.getPort();
+        boolean isHttps = "https".equalsIgnoreCase(uri.getScheme());
+        ProgressReporter progressReporter = (request.getContext() == null)
```

Comment:
we have a contract that `request.getContext()` is never null

Reply:
It being nullable is fine, and slightly preferable to me.